Bottlerocket Testing Cli (#487)
bottlerocket-os/bottlerocket-test-system: agent/test-agent-cli/design/RUNBOOK.md

```
Set alias   
```shell
alias cli="${TESTSYS_DIR}/.cargo/bin/cli"
```
Install the `cli` command line tool into the local CARGO_HOME as: 
   ```shell
   cd "${TESTSYS_DIR}"
   cargo install --path "${TESTSYS_DIR}/cli" --force
   ```

## Steps to create Bash based TestAgent:

The following commands can be used to communicate with a testsys cluster. Create a bash script like [Example test](../examples/example_test_agent_cli/example-test.sh).
```shell
# Get the configuration details and setting the task state running
test-agent-cli init

# Get the number of retires allowed in case of failing tests
test-agent-cli retry-count

# Get the secret value using the secret key
test-agent-cli get-secret secret-key

# Send the result of every test run to test object in Controller
test-agent-cli send-result -o pass -p 1 -f 0 -s 0

# Send any error encountered in test
test-agent-cli send-error error-message

# Mark the test as completed
test-agent-cli terminate --results-dir results_directory
```
Create a [Dockerfile](../examples/example_test_agent_cli/Dockerfile).\
Remember to set the ENTRYPOINT to the test Bash script and install the required packages.

> Assume everything is run from the root directory or use `TESTSYS_DIR` .

Reviewer: This was a comment for how the runbook commands should be set up, not something to add.

# Steps to use Test-agent Command Line Interface

In the current implementation of Test-agent, to create any test we have to implement the Runner trait in TestAgent and Client trait in Test agent takes care of sending and receiving the information from Kubernetes Cluster.

By this project, the command line program will take  place of the [TestAgent](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/5751fde5bf4d6ff11eb84c6e2eb70d36c1b0dee1/agent/test-agent/src/agent.rs#L27) object and the [Runner](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/5751fde5bf4d6ff11eb84c6e2eb70d36c1b0dee1/agent/test-agent/src/lib.rs#L51) trait implementation. 

The test object in Kubernetes  [controller](https://kubernetes.io/docs/concepts/architecture/controller/) will run the pod for Command Line Interface and this pod will report the results(that has been sent from user) to test object in Controller.

> This will enable the user to write the test using Bash Script and Command line Interface will help in receiving and sending the information from Kubernetes Cluster.

Reviewer: I'd start with:
''In addition to the test agent, testsys also provides users the ability to create bash tests using ...''

Reviewer: The higher level info will be in a generic testsys doc.

> In the current implementation of Test-agent, to create any test we have to implement the Runner trait in TestAgent and Client trait in Test agent takes care of sending and receiving the information from Kubernetes Cluster.
> 
> By this project, the command line program will take  place of the [TestAgent](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/5751fde5bf4d6ff11eb84c6e2eb70d36c1b0dee1/agent/test-agent/src/agent.rs#L27) object and the [Runner](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/5751fde5bf4d6ff11eb84c6e2eb70d36c1b0dee1/agent/test-agent/src/lib.rs#L51) trait implementation. 
> 
> The test object in Kubernetes  [controller](https://kubernetes.io/docs/concepts/architecture/controller/) will run the pod for Command Line Interface and this pod will report the results(that has been sent from user) to test object in Controller.
> 
> This will enable the user to write the test using Bash Script and Command line Interface will help in receiving and sending the information from Kubernetes Cluster.

Reviewer 2: I think we should simply this. I don't think we need the low-level description.

```suggestion
`test-agent-cli` lets you write Testsys test using Bash and help in receiving and sending information from/to the Testsys cluster.
```

## Important links :

* Source code for CLI: https://github.com/bottlerocket-os/bottlerocket-test-system/tree/develop/agent/test-agent-cli

> * Example test: https://github.com/bottlerocket-os/bottlerocket-test-system/tree/develop/agent/test-agent-cli/examples

Reviewer: These can be linked whenever they seem appropriate. I don't think it's necessary for them here.

## Prerequisites:

* kind: https://kind.sigs.k8s.io/docs/user/quick-start/
* Kubectl: https://kubernetes.io/docs/tasks/tools/
* Docker : https://docs.docker.com/get-started/
* Testsys Command line Interface

> ## Steps:

Reviewer: Should this be split into:
Create a bash based test agent (including image)
Use the bash based test agent

> * Testsys Command line Interface

Reviewer: Let's add a step for installing this and remove it as a prerequisite.

## Steps:

1. Create a Bash Test using the following commands to achieve the desired behavior of test.
```shell
# To get the configuration details and setting the task state running
test-agent-cli init

# To get the number of retires allowed in case of failing tests
test-agent-cli retry-count

# To send the result of every test run to test object in Controller
test-agent-cli send-result -o pass -p 1 -f 0 -s 0

# To get the secret value using the secret key
test-agent-cli get-secret secret-key

# To send any error encountered in test
test-agent-cli send-error error-message

# To create the Tar file of the results, Marking task_state as completed and handling keep_running behaviour
# Send the path and name of results directory where all the results have been saved
test-agent-cli terminate --results-dir results_directory
```

> 2. Create a docker file(to provide all the required resources to run the Bash script along with the Command Line Interface in a single pod), that will be used to create image that will run on the Test Pod.(See below example)

Reviewer: Can you link dockerfile to the one you created for the example agent.

> In this docker file make sure to mark entry point as the Test Bash script created in Step 1. Add the commands to install the packages required for Bashscript in docker file. 

Reviewer 2: nit:
```suggestion
In this Dockerfile make sure to set the `ENTRYPOINT` to the test Bash script created in Step 1. Add the commands to install the packages required for Bash script in the Dockerfile. 
```

In this docker file make sure to mark entry point as the Test Bash script created in Step 1. Add the commands to install the packages required for Bashscript in docker file. 

```shell
# syntax=docker/dockerfile:1.1.3-experimental

ARG BUILDER_IMAGE
FROM ${BUILDER_IMAGE} as build

ARG ARCH
USER root

# We need these environment variables set for building the `openssl-sys` crate

ENV PKG_CONFIG_PATH=/${ARCH}-bottlerocket-linux-musl/sys-root/usr/lib/pkgconfig
ENV PKG_CONFIG_ALLOW_CROSS=1
ENV CARGO_HOME=/src/.cargo
ENV OPENSSL_STATIC=true
ENV CARGO_HOME=/src/.cargo
ADD ./ /src/
WORKDIR /src/agent/test-agent-cli
RUN --mount=type=cache,mode=0777,target=/src/target 
cargo install --locked --offline --target ${ARCH}-bottlerocket-linux-musl --path .  --root ./

FROM public.ecr.aws/amazonlinux/amazonlinux:2
RUN yum update -y 
&& yum install -y jq 
&& yum clean all
COPY --from=build /src/agent/test-agent-cli/bin/test-agent-cli /usr/local/bin/
COPY --from=build /src/agent/test-agent-cli/examples/example_test_agent_cli/example-test.sh ./

ENTRYPOINT ["/bin/bash", "./example-test.sh"]
```

> 3. Create a docker image consists of the Command Line Interface, Bash Script test and all other resources required to run the test(execute following from Bottlerocket Test System  root folder).(Before running this make sure to add the details to create this image using docker build in [makefile](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/develop/Makefile))

Reviewer: Can we do this step without make. Just the docker build commands necessary.

   make example-test-agent-cli

> 4. Create a controller image(execute from Bottlerocket Test System root folder): `make controller`.

Reviewer: Let's put a step at the beginning to go to the root folder and then assume all calls are made from there.

> 5. Tag these images as eks as: `docker image tag controller controller:eks`

Reviewer: Let's use a tag that's relevant to what is happening.

> 6. Create any Resource Agent image(if current test depend on any Resource).

Reviewer: Let's keep this doc just about tests.

> 7. Create a kind cluster that will be used for Testsys `kind create cluster --name testsys`.

Reviewer: You could have a step for make sure this cluster doesn't exist and if it does here's how to delete it.

8. Load the docker image for Controller, resource and Test on Kind cluster as :
```shell

> kind load docker-image controller:eks --name cluster-name

Reviewer: We can load both the controller and test agent at the same time.

kind load docker-image example-test-agent-cli:eks --name testsys
```

> 8. Create the namespace, Test and Resource CRDs, roles for test and resource as `testsys install --controller-uri controller:eks`. The controller pod created will run the pod for Test:

Reviewer: We can just say install testsys to the cluster. I will have more in-depth testsys documentation coming.

> 9. Create the resource pod for the Resouce Agent on which the test will be dependent: `testsys run file filename`.

Reviewer: See above about resources.

10. Create a yaml file for Test pod as:
```shell
apiVersion: testsys.bottlerocket.aws/v1
kind: Test
metadata:
name: hello-world-cli
namespace: testsys-bottlerocket-aws
spec:
retries: 5
agent:
name: hello-agent-cli
image: example-test-agent-cli:eks
keepRunning: true
configuration:
mode: Fast
person: Bones the Cat
helloCount: 3
helloDurationMilliseconds: 25
resources: []

> dependsOn: []

Reviewer: Just link to example test agent yaml file.

```

> 11. Create the Test pod using the image(step 3) and yaml file (step 10) as `testsys run file filename`.

Reviewer: "Run the test by calling `testsys run file <filename>`"

> 12. Check the pod status using `kubectl get pods`, the Test pod status will be running if all the dependent resources and Test are available.  This Test pod will run the Bash Test created in step 1, and the output can be seen using `kubectl logs test-pod-name` or `testsys logs test-name` .

Reviewer: We can do all of this with just testsys so let's use it.

Reviewer: This documentation can be less high level because the testsys documentation will go into depth.

> 13. Once the test is complete, the Test Resource status will change to passed/failed/noTests  that can be checked using `testsys status -r -c` and Task State will change to Complete/Error that can be checked using `kubectl describe test` .

Reviewer: Let's remove Kubectl here.

14. Copy the Tar results file as `testsys results --n test-name --destination file-path`.

> 15. Set the keep_running status false(if has been set true in yaml file), that will change the pod status to complete.

Reviewer: How to cleanup?

> # Steps to use Test-agent Command Line Interface

Reviewer: Overall: Good work, can you run through each step and make them more command based, i.e 
1: Install kind (kind link)
2: Create a kind cluster that will be used for testsys: `kind create cluster --name agent-cli` 
...

Reviewer: Also make sure that the steps can be run through verbatim.

Author: I tried to make commands more direct now :)

In the current implementation of Test-agent, to create any test we have to implement the Runner trait in TestAgent and Client trait in Test agent takes care of sending and receiving the information from Kubernetes Cluster.

> By this project, the command line program will take  place of the [TestAgent](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/5751fde5bf4d6ff11eb84c6e2eb70d36c1b0dee1/agent/test-agent/src/agent.rs#L27) object and the [Runner](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/5751fde5bf4d6ff11eb84c6e2eb70d36c1b0dee1/agent/test-agent/src/lib.rs#L51) trait implementation. The test object in Kubernetes  [controller](https://kubernetes.io/docs/concepts/architecture/controller/) will run the pod for Command Line Interface and this pod will report the results(that has been sent from user) to test object in Controller.

Reviewer: nit: 1 line per sentence.

This will enable the user to write the test using Bash Script and Command line Interface will help in receiving and sending the information from Kubernetes Cluster.

## Important links :

* Source code for CLI: https://github.com/bottlerocket-os/bottlerocket-test-system/tree/develop/agent/test-agent-cli

> * Design document: https://quip-amazon.com/cEdxAuktp6j8

Reviewer: This link should be removed.

* Example test: https://github.com/bottlerocket-os/bottlerocket-test-system/tree/develop/agent/test-agent-cli/examples

## Prerequisites:

* kind: https://kind.sigs.k8s.io/docs/user/quick-start/
* Kubectl: https://kubernetes.io/docs/tasks/tools/
* Docker : https://docs.docker.com/get-started/

> * JQ or other tool to parse the Config JSON

Reviewer: Why does the user need jq?

Author: I just mentioned if user need any other tool to work with their bash script.

Reviewer: That doesn't need to be installed on their machine right?

Author: No, yeah that should be added in docker file, yeah make sense to remove it from here.

* Testsys Command line Interface

## Steps:

1. Create a bash Test that will consume the following commands to achieve the desired behavior

```shell
# To get the configuration details and setting the task state running
test-agent-cli init

# To get the number of retires allowed in case of failing tests
test-agent-cli retry-count

# To send the result of every test run to test object in Controller
test-agent-cli send-result -o pass -p 1 -f 0 -s 0

# To get the secret value using the secret key
test-agent-cli get-secret secret-key

# To send any error encountered in test
test-agent-cli send-error error-message

# To create the Tar file of the results, Marking task_state as completed and handling keep_running behaviour
# Send the path and name of results directory where all the results have been saved
test-agent-cli terminate --results-dir results_directory

# To mark the pod state as completed if keep_running has been set True in deploy.yaml file

> testsys set test-name --keep-running false

Reviewer: This doesn't belong in the bash script.

test-agent-cli init > ExampleConfig.json

echo "starting the Example test"

person=$(jq '.person' ExampleConfig.json)
helloCount=$(jq '.helloCount' ExampleConfig.json)
helloDurationMilliseconds=$(jq '.helloDurationMilliseconds' ExampleConfig.json)

echo "$person"
echo "$helloDurationMilliseconds"
echo "$helloCount"

# Get the number of retries allowed in case of a failing test
retries=$(test-agent-cli retry-count)
echo "no of retries are $retries"

# Get the secret for a key
secret=$(test-agent-cli get-secret key1)
echo "Value for secret is $secret"

# Perform the test with this bash script.

test () {
for ((i = 1; i <= helloCount; i++))
do
echo "Hello ${i} to ${person}"
sleep "${helloDurationMilliseconds}"
done
}

# Retry the test in case of failure
if ! test && [ "$retries" -gt 0 ]; then
for ((i = 1; i <= "$retries"; i ++))
do
echo "Test failed again, retrying ${i} of $retries"

# This will save the result of current iteration in a directory on kubernetes cluster,
# that will contain the info how many test cases failed, passed, skipped.
test-agent-cli send-result -o fail -p 0 -f 1 -s 0

if test; then

>     break

Reviewer: I don't think this needs to be in the docs it's already in the code base.

```shell
# syntax=docker/dockerfile:1.1.3-experimental

ARG BUILDER_IMAGE
FROM ${BUILDER_IMAGE} as build

ARG ARCH
USER root

# We need these environment variables set for building the `openssl-sys` crate

ENV PKG_CONFIG_PATH=/${ARCH}-bottlerocket-linux-musl/sys-root/usr/lib/pkgconfig
ENV PKG_CONFIG_ALLOW_CROSS=1
ENV CARGO_HOME=/src/.cargo
ENV OPENSSL_STATIC=true
ENV CARGO_HOME=/src/.cargo
ADD ./ /src/
WORKDIR /src/agent/test-agent-cli
RUN --mount=type=cache,mode=0777,target=/src/target 
cargo install --locked --offline --target ${ARCH}-bottlerocket-linux-musl --path .  --root ./

FROM public.ecr.aws/amazonlinux/amazonlinux:2
RUN yum update -y 
&& yum install -y jq 
&& yum clean all
COPY --from=build /src/agent/test-agent-cli/bin/test-agent-cli /usr/local/bin/
COPY --from=build /src/agent/test-agent-cli/examples/example_test_agent_cli/example-test.sh ./

ENTRYPOINT ["/bin/bash", "./example-test.sh"]

> ```

Reviewer: Can you explain what is boilerplate and what is configurable.

Author: I have added that whatever they need to run Bash script, they can add that in docker file. Should I mention that The way to copy test agent cli, example test and other part of the file should remain same?

Reviewer: Let's remove the docker file code here and make sure the example dockerfile is well documented.

> 3. Create a docker image consists of the Command Line Interface, Bash Script test and all other resources required to run the test as running the following command from Bottlerocket Test System root folder.(Before running this make sure to add the details to create this image using docker build in makefile(https://github.com/bottlerocket-os/bottlerocket-test-system/blob/develop/Makefile))

Reviewer: Can you make `makefile` a link to our makefile.

   make example-test-agent-cli
4. Create a controller image using the following command from Bottlerocket Test System root folder.
   `make controlller`
5. Create any Resource Agent image(if current test depend on any Resource) in the same way as we do in existing process.

> 6. Create a kind cluster as *kind create cluster --name testsys*. I prefer to first delete the cluster if any exists using *kind delete cluster --name testsys* .

Reviewer: We can use "`" instead of *

Reviewer: I'd remove "I prefer to first delete the cluster if any exists using *kind delete cluster --name testsys* ."

7. Load the docker image for Controller, resource and Test on Kind cluster as :

```shell

> kind load docker-image controller:eks --name cluster-name

Reviewer: Where are these docker images coming from?

Author: This I have explained in step 4 how to create these images.

Reviewer: There is no place in the docs to tag the images.

kind load docker-image example-test-agent-cli:eks --name testsys
```

> 8. Run the Testsys install that will create the namespace, Test and Resource CRDs, roles for test and resource and many more listed in https://quip-amazon.com/aJbvA3bH0WNO#temp:C:UIBdf4817383b17480fb48010c60 .

Reviewer: There shouldn't be links to docs out of GitHub.

   `testsys install --controller-uri controller:eks`
   The controller pod created will run the pod for Command Line Interface and Test.

> 9. Create the resource pod for the Resouce Agent on which the test will be dependent using kubectl apply -f deploy.yaml ** or testsys run file filename.

Reviewer: Let's use `testsys` instead of `Kubectl`

10. Now create a yaml file for Test pod as:

```shell
apiVersion: testsys.bottlerocket.aws/v1
kind: Test
metadata:
name: hello-world-cli
namespace: testsys-bottlerocket-aws
spec:
retries: 5
agent:
name: hello-agent-cli
image: example-test-agent-cli:eks
keepRunning: true

> timeout: "500m"

Reviewer: Timeout is optional and should probably not be here.

# Steps to use Test-agent Command Line Interface
`test-agent-cli` lets you write Testsys test using Bash and help in receiving and sending information from/to the Testsys cluster.

> ## Prerequisites:

Reviewer: Since the testsys cli is going to be removed eventually lets switch over to cli instead.

Reviewer: I would also remove the step numbers.

* kind: https://kind.sigs.k8s.io/docs/user/quick-start/
* Kubectl: https://kubernetes.io/docs/tasks/tools/
* Docker : https://docs.docker.com/get-started/

## Steps to install Testsys
1. Set the TESTSYS_DIR variable to point to the directory in which you have cloned the project. For example:`export TESTSYS_DIR="${HOME}/repos/bottlerocket-test-system"`
2. Set alias `alias testsys="${TESTSYS_DIR}/.cargo/bin/testsys"`
3. Install the testsys command line tool into the local CARGO_HOME as: 
```shell
cd "${TESTSYS_DIR}"
cargo install --path "${TESTSYS_DIR}/bottlerocket/testsys" --force

> ```

Reviewer: `cargo install --path cli`

## Steps to create Bash based TestAgent:

> 1. Create a Bash Test using the following commands to achieve the desired behavior of test.

Reviewer: The following commands can be used to communicate with a testsys cluster. Create a bash script like ...

```shell

> # To get the configuration details and setting the task state running

Reviewer: nit: Get the configuration details and setting the task state running

test-agent-cli init

# To get the number of retires allowed in case of failing tests
test-agent-cli retry-count

# To send the result of every test run to test object in Controller
test-agent-cli send-result -o pass -p 1 -f 0 -s 0

# To get the secret value using the secret key
test-agent-cli get-secret secret-key

# To send any error encountered in test
test-agent-cli send-error error-message

# To create the Tar file of the results, Marking task_state as completed and handling keep_running behaviour
# Send the path and name of results directory where all the results have been saved
test-agent-cli terminate --results-dir results_directory
```

> 2. Create a docker file(to provide all the required resources to run the Bash script along with the Command Line Interface in a single pod), that will be used to create image that will run on the Test Pod. Example Dockerfile : [Example Agent Dockerfile](../examples/example_test_agent_cli/Dockerfile).\

Reviewer: Create a [Dockerfile](../examples/example_test_agent_cli/Dockerfile)

>    In this Dockerfile make sure to set the `ENTRYPOINT` to the test Bash script created in Step 1. Add the commands to install the packages required for Bash script in the Dockerfile.

Reviewer: Remember to set the [`ENTRYPOINT`](<link to entry point line number>) to the test Bash script and install the required packages.

> 3. Move to [root](https://github.com/bottlerocket-os/bottlerocket-test-system) of the repository.

Reviewer: Assume everything is run from the root directory or use `TESTSYS_DIR`.

> 4. Create a docker image consists of the Command Line Interface, Bash Script test and all other resources required to run the test: 

Reviewer: 4. Create the docker image.
script (image can be tagged here in the build command)

   ```shell 
   docker build example-test-agent-cli 
   -f agent/test-agent-cli/examples/example_test_agent_cli/Dockerfile
   ``` 
   and tag this image as : `docker image tag example-test-agent-cli example-test-agent-cli:bash`

> 5. Create a controller image: `make controller` and tag this image as bash as: `docker image tag controller controller:bash`.

Reviewer: Create and tag the controller image.
```bash
make controller
docker tag controller controller:bash
```

## Steps to use Bash based TestAgent:

> 1. Check if the cluster already exists: `kind get clusters`, if it exists delete it as: 

Reviewer: Check if the kind cluster already exists.
```bash
kind get clusters
```

If the cluster already exists, it should be deleted.
```bash
kind delete cluster --name <whatever name is used in the demo>
```

   ```shell
   kind delete cluster --name <testsys_cluster_name>
   ```

>    Create a kind cluster that will be used for Testsys `kind create cluster --name <testsys_cluster_name>`.

Reviewer: Create the new cluster.
```bash
kind create cluster --name <whatever name is used in the demo>
```

2. Load the docker image for Controller, resource and Test on Kind cluster as :
   ```shell
   kind load docker-image controller:bash, example-test-agent-cli:bash --name <testsys_cluster_name>

>    ```

Reviewer: Now the images created earlier need to be added to the cluster.
```bash
kind load docker-image controller:demo example-test-agent:demo --name testsys-demo
```

(change the names to match)

> 3. Install Testsys to the cluster as:  

Reviewer: Install Testsys to the cluster.

   ```shell
   testsys install --controller-uri controller:bash 
   ```

>    The controller pod created will run the pod for Test.

Reviewer: This is covered in other docs.

> 4. Create a yaml file for Test pod as: [Example yaml file](../tests/data/deploy_test.yaml)

Reviewer: Create a [yaml file](link) for test.

> 5. Run the test by calling `testsys run file <filename>`.

Reviewer: Run the test
```bash
command
```

> 6. Check the resource status using `testsys status -c -r`. This Test pod will run the Bash Test created in step 1, and the output can be seen using `testsys logs <test-name>`.

Reviewer: Check the status of the test.
```bash
command
```

7. Once the test is complete, the Test Resource status will change to passed/failed/noTests  that can be checked as
   ```shell 
   testsys status -r -c
   ```

>    and Task State will change to Complete/Error.

Reviewer: This step will be covered in other docs.

> 8. Copy the Tar results file as `testsys results --n test-name --destination file-path`.

Reviewer: This will be covered in other docs.

9. Set the keep_running status false(if has been set true in yaml file), that will change the pod status to complete as:
   ```shell
   testsys set test-name --keep-running false 

>    ```

Reviewer: This step isn't necessary.

```
Set alias   
```shell
alias cli="${TESTSYS_DIR}/.cargo/bin/cli"
```
Install the `cli` command line tool into the local CARGO_HOME as: 
   ```shell
   cd "${TESTSYS_DIR}"
   cargo install --path "${TESTSYS_DIR}/cli" --force
   ```

## Steps to create Bash based TestAgent:

The following commands can be used to communicate with a testsys cluster. Create a bash script like [Example test](../examples/example_test_agent_cli/example-test.sh).
```shell
# Get the configuration details and setting the task state running
test-agent-cli init

# Get the number of retires allowed in case of failing tests
test-agent-cli retry-count

# Get the secret value using the secret key
test-agent-cli get-secret secret-key

# Send the result of every test run to test object in Controller
test-agent-cli send-result -o pass -p 1 -f 0 -s 0

# Send any error encountered in test
test-agent-cli send-error error-message

# Mark the test as completed
test-agent-cli terminate --results-dir results_directory
```
Create a [Dockerfile](../examples/example_test_agent_cli/Dockerfile).\
Remember to set the ENTRYPOINT to the test Bash script and install the required packages.

Create the docker image.
   ```shell 

>    make example-test-agent-cli 

Reviewer: Can you add a note that users should add a target to the makefile if they want to create their new image.

> # Steps to use Test-agent Command Line Interface

Reviewer 3: Yesterday we merged a new `make mdlint` target that performs markdown linting. There are a bunch of things here that the linting is going to flag. Can you rebase on the latest from `develop` and run the mdlint target to help clean up things. They are all very minor formatting things.

Reviewer 3: Oops, I thought it was merged, but still pending:

https://github.com/bottlerocket-os/bottlerocket-test-system/pull/514

So if you would like to run that command to get the list of issues, that would be great. Otherwise if this merges first, I can update that PR to clean things up. Either way is fine with me. :)

Author: As I had discussion with @stmcginnis, the cleanup will be done in the [#514](https://github.com/bottlerocket-os/bottlerocket-test-system/pull/514)

> `test-agent-cli` lets you write Testsys test using Bash and help in receiving and sending information from/to the Testsys cluster.

Reviewer 2: It seems like we use "TestSys" instead of "Testsys" as the proper capitalization.

Applies elsewhere as well.

## Prerequisites:
* kind: https://kind.sigs.k8s.io/docs/user/quick-start/
* Kubectl: https://kubernetes.io/docs/tasks/tools/
* Docker : https://docs.docker.com/get-started/

## Steps to install Testsys
Set the TESTSYS_DIR variable to point to the directory in which you have cloned the project. For example:
```shell
export TESTSYS_DIR="${HOME}/repos/bottlerocket-test-system"
```
Set alias   
```shell
alias cli="${TESTSYS_DIR}/.cargo/bin/cli"
```
Install the `cli` command line tool into the local CARGO_HOME as: 
   ```shell
   cd "${TESTSYS_DIR}"
   cargo install --path "${TESTSYS_DIR}/cli" --force
   ```

## Steps to create Bash based TestAgent:

The following commands can be used to communicate with a testsys cluster. Create a bash script like [Example test](../examples/example_test_agent_cli/example-test.sh).
```shell

> # Get the configuration details and setting the task state running

Reviewer 2: ```suggestion
# Get the configuration details and set the task state running
```
